### PR TITLE
Add node connection manager

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Introduces a new `ClientManager` for managing gRPC client connections
- Implements a `RegistrationFunc` callback for registering gRPC services
- Refactors `NewAPIServer` to use the `RegistrationFunc` instead of direct service creation
- Adds `BuildClient` method to `Node` struct for creating gRPC client connections
- Updates test utilities to use the new service registration approach

## Why a registration func?

Two reasons:
1. I wanted to get the dependencies and logic of building the individual services out of the `api` module and make that solely focused on creating an API server
2. Some server's will have just a payer API, some will have a message and payer API. I wanted to move that decision making up a layer in the stack.

## AI Assisted Notes

This PR introduces a `ClientManager` to handle gRPC client connections efficiently. It implements a `RegistrationFunc` callback for registering gRPC services, allowing for more flexible service configuration. The `NewAPIServer` function has been refactored to use this new `RegistrationFunc` instead of creating services directly.

A `BuildClient` method has been added to the `Node` struct, simplifying the process of creating gRPC client connections. Test utilities have been updated to reflect these changes and use the new service registration approach.

These changes improve the modularity and flexibility of the API server setup and client connection management.

### CODE REVIEW

**Comment by neekolas** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/243?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#243** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/243?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#241** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/241?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#239** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/239?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#238** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/238?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#237** <a href="https://app.graphite.dev/github/pr/xmtp/xmtpd/237?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @neekolas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: pkg/api/payer/clientManager.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	if ok {
+		return existing.(*grpc.ClientConn), nil
+	}
+
+	conn, err := c.newClientConnection(nodeID)
+	if err != nil {
+		return nil, err
+	}
+	// Store the connection
+	c.connections.Store(nodeID, conn)
```
Comment (richardhuaaa) [inline pkg/api/payer/clientManager.go:40]: I noticed that we're using a `sync.Map` here, implying some degree of concurrency. Just noting that we seem to have a `write-after-read` race condition here, which may or may not matter depending on what you're trying to use this for

**Review by richardhuaaa — APPROVED** [review]
Love the focus on clean abstractions and separation of concerns

File: pkg/api/payer/clientManager.go (hunk 1 above)
Comment (neekolas) [reply pkg/api/payer/clientManager.go]: It shouldn't really matter, but it's easy enough to fix so that it's bulletproof. Maybe will prevent some future error, so I updated it.

**Comment by neekolas** [comment]
### Merge activity

* **Oct 21, 3:48 PM PDT**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/xmtp/xmtpd/243).
